[16.0][ADD] stock_release_channel_delivery

glitchov opened the pull request:
This addon adds a field containing the carrier linked to the release channel. 
Though an overlap can be seen on the picking point of view as the
rule_domain does the job the idea is to easy link the channel with the
customer.

glitchov committed:
[16.0][ADD] stock_release_channel_delivery

sbejaoui committed:
[IMP] - stock_release_channel_delivery: better handle carrier filter

jbaudoux commented:
/ocabot rebase

jbaudoux commented:
@sbejaoui Can you rebase, it seems failing tests are not related to this PR

sbejaoui commented:
> @sbejaoui Can you rebase, it seems failing tests are not related to this PR

done

jbaudoux commented on stock_release_channel_delivery/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "Stock Release Channel Delivery",
+    "summary": """
+        Exposes clearly the carrier linked to the release channel """,
```
```suggestion
        Add a carrier selection criteria on the release channel """,
```

jbaudoux commented on stock_release_channel_delivery/models/stock_picking.py:
```diff
@@ -0,0 +1,21 @@
+
+    def _get_release_channel_possible_candidate_domain(self):
+        domain = super()._get_release_channel_possible_candidate_domain()
+        if self.carrier_id:
```
```suggestion
        if self.carrier_ids:
```

jbaudoux commented on stock_release_channel_delivery/models/stock_picking.py:
```diff
@@ -0,0 +1,21 @@
+            domain.extend(
+                [
+                    "|",
+                    ("carrier_id", "=", False),
+                    ("carrier_id", "=", self.carrier_id.id),
```
```suggestion
                    ("carrier_ids", "=", False),
                    ("carrier_ids", "in", self.carrier_id.ids),
```

jbaudoux commented on stock_release_channel_delivery/models/stock_release_channel.py:
```diff
@@ -0,0 +1,15 @@
+
+    _inherit = "stock.release.channel"
+
+    carrier_id = fields.Many2one(
```
```suggestion
    carrier_ids = fields.Many2many(
```

jbaudoux commented on stock_release_channel_delivery/models/stock_release_channel.py:
```diff
@@ -0,0 +1,15 @@
+
+    carrier_id = fields.Many2one(
+        comodel_name="delivery.carrier",
+        help="The carrier linked to the release channel.",
+        string="Transporter",
```
```suggestion
        string="Shipping Methods",
```

jbaudoux commented on stock_release_channel_delivery/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,4 @@
+This addon adds a field containing the id of the carrier linked to the release
+channel. Though an overlap can be seen on the picking point of view as the
+rule_domain does the job the idea is to easy link the channel with the
+customer.
```
```suggestion
This addon adds a selection criteria on the release channel based on the shipping method.
When you don't classify your deliveries in different operation types per shipping method, then you can use this criteria to filter the selection of valid release channels. While you could achieve the same with the domain provided on the base release channel module, this criteria provides a more explicit configuration and also performs better.
```

jbaudoux commented on stock_release_channel_delivery/views/stock_release_channel_views.xml:
```diff
@@ -0,0 +1,20 @@
+        />
+        <field name="arch" type="xml">
+            <field name="rule_domain" position="before">
+                <field name="carrier_id" />
```
```suggestion
                <field name="carrier_ids" widget="many2many_tags"/>
```

jbaudoux commented on stock_release_channel_delivery/tests/test_release_channel_delivery.py:
```diff
@@ -0,0 +1,43 @@
+        picking and release channel have the same carrier
+        """
+        self.picking.carrier_id = self.carrier
+        self.channel.carrier_id = self.carrier
```
```suggestion
        self.channel.carrier_ids = self.carrier
```

jbaudoux commented on stock_release_channel_delivery/tests/test_release_channel_delivery.py:
```diff
@@ -0,0 +1,43 @@
+        picking have carrier but not the release channel
+        """
+        self.picking.carrier_id = self.carrier
+        self.channel.carrier_id = False
```
```suggestion
        self.channel.carrier_ids = False
```

jbaudoux commented on stock_release_channel_delivery/tests/test_release_channel_delivery.py:
```diff
@@ -0,0 +1,43 @@
+        picking and release channel have different carrier
+        """
+        self.picking.carrier_id = self.carrier
+        self.channel.carrier_id = self.carrier2
```
```suggestion
        self.channel.carrier_ids = self.carrier2
```

jbaudoux commented on stock_release_channel_delivery/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+        Exposes clearly the carrier linked to the release channel """,
+    "version": "16.0.1.0.1",
+    "license": "AGPL-3",
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
```
```suggestion
    "author": "ACSONE SA/NV,BCIM,Odoo Community Association (OCA)",
```

jbaudoux commented on stock_release_channel_delivery/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1 @@
+* Hughes Damry <hughes.damry@acsone.eu>
```
```suggestion
* Hughes Damry <hughes.damry@acsone.eu>
* Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

jbaudoux reviewed (changes requested):
We cannot be so restrictive to only one shipping method on the release channel.

In odoo, there are some providers requiring to create one shipping method for national deliveries and another one for international deliveries. However they are usually managed the same way in the warehouse and classified on the same release channel.

simahawk commented:
@glitchov are you planning to move fwd soon here? Otherwise we could take over.

simahawk commented:
ping :)

lmignon commented:
@simahawk We are stuck on stabilizing our Odoo 16 instance.... We've no time to manage the change and the migration of a `Many2one` to `Many2many` field type for the carrier on the release channel model.

lmignon committed:
[FIX] stock_release_channel_delivery: No assignation if carriers differ

A picking without carrier should not be assigned to a release channel restricted to a specific carrier.

lmignon committed:
[IMP] stock_release_channel_delivery: Add missing contributors

lmignon committed:
Update stock_release_channel_delivery/models/stock_release_channel.py

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

jbaudoux commented:
replaced by https://github.com/OCA/wms/pull/840
